gma/src/android/ump/consent_info_internal_android.cc

```diff
@@ -0,0 +1,667 @@
+
+  ReleaseClasses(env);
+  util::Terminate(env);
+
+  env->DeleteGlobalRef(activity_);
+  activity_ = nullptr;
+  java_vm_ = nullptr;
+}
+
+// clang-format off
+#define ENUM_VALUE(class_namespace, field_name)           \
+   env->GetStaticIntField(class_namespace::GetClass(),    \
+                          class_namespace::GetFieldId(class_namespace::k##field_name))
+// clang-format on
+
+void ConsentInfoInternalAndroid::CacheEnumValues(JNIEnv* env) {
+  // Cache enum values when the class loads, to avoid JNI lookups during
+  // callbacks later on when converting enums between Android and C++ values.
+  enums_.consentstatus_unknown =
+      ENUM_VALUE(consentinformation_consentstatus, Unknown);
+  enums_.consentstatus_required =
+      ENUM_VALUE(consentinformation_consentstatus, Required);
+  enums_.consentstatus_not_required =
+      ENUM_VALUE(consentinformation_consentstatus, NotRequired);
+  enums_.consentstatus_obtained =
+      ENUM_VALUE(consentinformation_consentstatus, Obtained);
+
+  enums_.debug_geography_disabled =
+      ENUM_VALUE(consentdebugsettings_debuggeography, Disabled);
+  enums_.debug_geography_eea =
+      ENUM_VALUE(consentdebugsettings_debuggeography, EEA);
+  enums_.debug_geography_not_eea =
+      ENUM_VALUE(consentdebugsettings_debuggeography, NotEEA);
+
+  enums_.formerror_success = 0;
+  enums_.formerror_internal = ENUM_VALUE(formerror_errorcode, InternalError);
+  enums_.formerror_network = ENUM_VALUE(formerror_errorcode, InternetError);
+  enums_.formerror_invalid_operation =
+      ENUM_VALUE(formerror_errorcode, InvalidOperation);
+  enums_.formerror_timeout = ENUM_VALUE(formerror_errorcode, TimeOut);
+
+  enums_.privacy_options_requirement_unknown =
+      ENUM_VALUE(consent_info_helper, PrivacyOptionsRequirementUnknown);
+  enums_.privacy_options_requirement_required =
+      ENUM_VALUE(consent_info_helper, PrivacyOptionsRequirementRequired);
+  enums_.privacy_options_requirement_not_required =
+      ENUM_VALUE(consent_info_helper, PrivacyOptionsRequirementNotRequired);
+
+  enums_.function_request_consent_info_update =
+      ENUM_VALUE(consent_info_helper, FunctionRequestConsentInfoUpdate);
+  enums_.function_load_consent_form =
+      ENUM_VALUE(consent_info_helper, FunctionLoadConsentForm);
+  enums_.function_show_consent_form =
+      ENUM_VALUE(consent_info_helper, FunctionShowConsentForm);
+  enums_.function_load_and_show_consent_form_if_required =
+      ENUM_VALUE(consent_info_helper, FunctionLoadAndShowConsentFormIfRequired);
+  enums_.function_show_privacy_options_form =
+      ENUM_VALUE(consent_info_helper, FunctionShowPrivacyOptionsForm);
+  enums_.function_count = ENUM_VALUE(consent_info_helper, FunctionCount);
+}
+
+void ConsentInfoInternalAndroid::JNI_ConsentInfoHelper_completeFuture(
+    JNIEnv* env, jclass clazz, jint future_fn, jlong consent_info_internal_ptr,
+    jlong future_handle, jint error_code, jobject error_message_obj) {
+  MutexLock lock(s_instance_mutex);
```

Comment:
I think there might be a problem with the locks.  If the destructor is called, it gets the lock on s_instance_mutex, and then calls disconnect, which locks the java object.  In some of the java calls, it locks the java object, and then calls completeFuture, which locks s_instance_mutex.

Reply:
Yeah, sounds like this could cause a deadlock, good catch. I think if I do the Disconnect first, prior to locking s_instance, it should be safe.